src/Compilers/CSharp/Test/Emit/PDB/PDBTests.cs

```diff
@@ -9731,6 +9731,453 @@ .maxstack  2
+        <entry offset=""0x3f"" hidden=""true"" document=""1"" />
+        <entry offset=""0x43"" startLine=""15"" startColumn=""5"" endLine=""15"" endColumn=""6"" document=""1"" />
+      </sequencePoints>
+      <scope startOffset=""0x0"" endOffset=""0x45"">
+        <namespace name=""System"" />
+        <scope startOffset=""0x1"" endOffset=""0x43"">
+          <local name=""CS$&lt;&gt;8__locals0"" il_index=""0"" il_start=""0x1"" il_end=""0x43"" attributes=""0"" />
+        </scope>
+      </scope>
+    </method>
+    <method containingType=""C+&lt;&gt;c__DisplayClass0_0"" name=""&lt;F&gt;b__0"">
+      <customDebugInfo>
+        <forward declaringType=""C"" methodName=""F"" parameterNames=""o"" />
+        <encLocalSlotMap>
+          <slot kind=""28"" offset=""86"" />
+          <slot kind=""temp"" />
+        </encLocalSlotMap>
+      </customDebugInfo>
+      <sequencePoints>
+        <entry offset=""0x0"" startLine=""8"" startColumn=""42"" endLine=""12"" endColumn=""14"" document=""1"" />
+        <entry offset=""0xa"" startLine=""8"" startColumn=""48"" endLine=""12"" endColumn=""14"" document=""1"" />
+        <entry offset=""0xb"" hidden=""true"" document=""1"" />
+        <entry offset=""0x16"" startLine=""10"" startColumn=""22"" endLine=""10"" endColumn=""23"" document=""1"" />
+        <entry offset=""0x1a"" startLine=""11"" startColumn=""22"" endLine=""11"" endColumn=""23"" document=""1"" />
+        <entry offset=""0x1e"" hidden=""true"" document=""1"" />
+        <entry offset=""0x21"" startLine=""8"" startColumn=""42"" endLine=""12"" endColumn=""14"" document=""1"" />
+        <entry offset=""0x22"" hidden=""true"" document=""1"" />
+      </sequencePoints>
+    </method>
+  </methods>
+</symbols>");
+        }
+
+        [WorkItem(50321, "https://github.com/dotnet/roslyn/issues/50321")]
+        [ConditionalFact(typeof(CoreClrOnly))]
+        public void NestedSwitchExpressions_Closures_02()
+        {
+            string source = WithWindowsLineBreaks(
+@"using System;
+class C
+{
+    static string F(object o)
+    {
+        return o switch
+        {
+            int i => new Func<string>(() => ""1"" + i switch
+            {
+                1 => new Func<string>(() => ""2"" + i)(),
+                _ => ""3""
+            })(),
+            _ => ""4""
+        };
+    }
+}");
+            var verifier = CompileAndVerify(source, options: TestOptions.DebugDll);
+            verifier.VerifyTypeIL("C",
+@".class private auto ansi beforefieldinit C
+	extends [netstandard]System.Object
+{
+	// Nested Types
+	.class nested private auto ansi sealed beforefieldinit '<>c__DisplayClass0_0'
+		extends [netstandard]System.Object
+	{
+		.custom instance void [netstandard]System.Runtime.CompilerServices.CompilerGeneratedAttribute::.ctor() = (
+			01 00 00 00
+		)
+		// Fields
+		.field public int32 '<i>5__2'
+		.field public class [netstandard]System.Func`1<string> '<>9__1'
```

Comment:
>.field public class [netstandard]System.Func`1<string> '<>9__1' [](http://example.com/codeflow?start=2&length=63)

There is one difference to the generated code with this change. The delegate for the nested lambda is cached in this instance field on the closure class. This was not cached previously was because `ClosureConversion.RewriteLambdaConversion()` did not recognize this as a nested lambda - the containing scope passed to `InLoopOrLambda()` was `BoundBlock` with the incorrect syntax so `InLoopOrLambda()` returned `false`.